src/peft/tuners/tuners_utils.py

```diff
@@ -433,6 +433,21 @@ def inject_adapter(self, model: nn.Module, adapter_name: str, autocast_adapter_d
         # update peft_config.target_modules if required
         peft_config = _maybe_include_all_linear_layers(peft_config, model)
 
+        # This is an optimization to reduce the number of entries in the target_modules list. The reason is that in some
+        # circumstances, target_modules can contain hundreds of entries. Since each target module is checked against
+        # each module of the net (which can be thousands), this can become quite expensive when many adapters are being
+        # added. Often, the target_modules can be condensed in such a case, which speeds up the process.
+        # A context in which this can happen is when diffusers loads non-PEFT LoRAs. As there is no meta info on
+        # target_modules in that case, they are just inferred by listing all keys from the state_dict, which can be
+        # quite a lot. See: https://github.com/huggingface/diffusers/issues/9297
+        # As there is a small chance for undiscovered bugs, we apply this optimization only if the list of
+        # target_modules is sufficiently big.
+        if isinstance(peft_config.target_modules, (list, set)) and len(peft_config.target_modules) >= 20:
+            names_not_match = [n for n in key_list if n not in peft_config.target_modules]
+            new_target_modules = find_minimal_target_modules(peft_config.target_modules, names_not_match)
```

Comment:
Maybe we keep `find_minimal_target_modules()` as a pseudo-private method to denote the experimental nature of this feature?